classes TriaxialEllipsoid, ProlateEllipsoid and OblateEllipsoid

birocoles opened the pull request:
This PR creates three new mesher classes to deal with ellipsoidal models. 
The classes are called TriaxialEllipsoid, ProlateEllipsoid and OblateEllipsoid,
and are used to create triaxial, prolate and oblate ellipsoids, respectively.
Such classes would be useful for modelling the gravity and magnetic fields
produced by ellipsoids.

### Checklist:

- [x] Make tests for new code (at least 80% coverage)
- [x] Create/update docstrings
- [x] Include relevant equations and citations in docstrings
- [x] Docstrings follow the style conventions
- [x] Code follows PEP8 style conventions
- [x] Code and docs have been spellchecked
- [x] Include new dependencies in `doc/install.rst`, `environment.yml`, `ci/requirements-conda.txt` and `ci/requirements-pip.txt`.
- [x] Documentation builds properly (run `cd doc; make` locally)
- [ ] Changelog entry (leave for last to avoid conflicts)
- [x] First-time contributor? Add yourself to `doc/contributors.rst` (leave for last to avoid conflicts)

birocoles committed:
create the masher classes TriaxialEllipsoid, ProlateEllipsoid and OblateEllipsoid

birocoles committed:
remove the new gravmag files and correct pep8 problems in mesher.py

birocoles committed:
correct pep8 problems

birocoles committed:
correct pep8 problems

birocoles committed:
correct pep8 problems

birocoles committed:
correct some doctests and pep8 problems

birocoles commented:
@leouieda , I have no idea about the problem with AppVeyor.

birocoles commented:
@leuieda, I guess I have just found the problem. Apparently, there is an assertion error in the test_eqlayer.py.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+
+
+class TriaxialEllipsoid (GeometricElement):
+
```
Watch out for these extra white spaces before the `(` and the extra newline between a function/class and the docstring. They appear in a lot of places below. Just remove them.

birocoles replied:
Done

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+
+    """
+    Create a triaxial ellipsoid.
+
```
Add some more sentences below this one to briefly explain what is a triaxial ellipsoid and what are the axis and angles you define below.

birocoles replied:
Done

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+    * x, y, z : float
+        The coordinates of the center of the ellipsoid
+    * a, b, c : float
+        Semi-axes of the ellipsoid (a>b>c)
```
`a, b, c` are terrible variable names. It's really hard to know what they are just from the name and they mean different things for different ellipsoids. It would be better to use their proper names if they have them. If not, use `large_axis`, `median_axis`, `small_axis` or something along those lines.

birocoles replied:
Done

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+    * a, b, c : float
+        Semi-axes of the ellipsoid (a>b>c)
+    * alpha, gamma, delta
+        Orientation angles of the ellipsoid
```
Same as the comment about the axes names. Use something more informative like dip, inclination, azimuth, etc.

birocoles replied:
Done

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+    * props : dict
+        Physical properties assigned to the ellipsoid.
+        Ex: ``props={'density':10,
+                     'remanence':[10000., 25., 40.],
```
Is `remanence` the remanent magnetization vector? If so, then `remanent magnetization` is a better name.

birocoles replied:
Yes, `remanence` means remanent magnetization vector. I replaced `remanence` by `remanent magnetization`.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+        Physical properties assigned to the ellipsoid.
+        Ex: ``props={'density':10,
+                     'remanence':[10000., 25., 40.],
+                     'k':[0.562, 0.485, 0.25, 90., 0., 0.]}``
```
What is `k`?

birocoles replied:
It is complicated task. `k` is the susceptibility tensor. I have replaced `k` by `susceptibility tensor`. The susceptibility tensor is defined as follows (in Latex):
$$
\mathbf{K} = \mathbf{U} \left[
\begin{array}{ccc}
k_{1} & & \\
& k_{2} & \\
& & k_{3} \\
\end{array}
\right] \mathbf{U}^{\top} \: ,
$$
where $k_{1}$, $k_{2}$ and $k_{3}$ are the *principal susceptibilities* and $\mathbf{U}$ is an orthogonal matrix whose columns are defined by unit vectors $\mathbf{u}_{1}$, $\mathbf{u}_{2}$ and $\mathbf{u}_{3}$. These vectors are called *principal directions*. We define the principal directions by using three angles: strike, rake and dip. These angles have the same geometrical meaning as those used to define the spatial orientation of the ellipsoid, but they are different.

We define the susceptibility tensor as a list containing the principal susceptibilities and the three angles. I am not sure if this is the best way of defining the susceptibility. This way can be used to define any (I am not sure) matrix. Notice that the principal susceptibilities represent the eigenvalues and $\mathbf{U}$ is the matrix containing the eigenvectors of the susceptibility tensor. Maybe, we could create a function similar to the `utils.ang2vec`. This function would receive the eigenvalues and three angles defining the eigenvectors and return the resultant matrix. What do you think?

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+        >>> e.props['remanence']
+        [10000.0, 25.0, 40.0]
+        >>> e.addprop('density', 20)
+        >>> print e.props['density']
```
Avoid `print` statements in doctests. In Python 3, `print` is a function. You can use `print(e.props['density'])` in the doctests and it should work. This will make the transition to Python 3 easier.

birocoles replied:
I replaced `print` by `print(...)` in the doctests.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+
+    Examples:
+
+        >>> e = TriaxialEllipsoid(1, 2, 3, 6, 5, 4, 10, 20, 30,{
```
When creating the instance, use the argument names to make it easier to understand. There is no way that I can know what `1, 2, 3, 6, 5, 4, 10, 20, 30` means for the ellipsoid. Use `TriaxialEllipsoid(x=1,0 y=2, z=3, a=6, b=5, c=4, alpha=10, beta=20, gamma=30, props={...`.

birocoles replied:
Done

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+        self.c = float(c)
+        self.alpha = float(alpha)
+        self.gamma = float(gamma)
+        self.delta = float(delta)
```
Why all the `float` conversions?

birocoles replied:
I removed the float conversions.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+        names = names + [(p, self.props[p]) for p in sorted(self.props)]
+        return ' | '.join('%s:%s' % (n, v) for n, v in names)
+
+    def V(self, angles=None):
```
Please change the name from `V` to something more informative, like `coord_trans_matrix` or something like that.

leouieda replied:
Also, why `angles` and not `alpha, gamma, delta`? That would eliminate the `assert len` check below and be more explicit (you don't have to remember which angle comes first).

birocoles replied:
I removed the method `def V(...)`.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+
+        * angles: array like containing the three
+            orientation angles (alpha, gamma, delta) of the main
+            susceptibility axes.
```
So it can depend on the susceptibility? Does it have to be? Couldn't this be just some other directional vector and then the magnetic modeling functions pass in the susceptibility? This is a matrix that converts from a given direction to the ellipsoid axis direction right?

leouieda replied:
Also, need to explain what happens when the values aren't given (will use the angles of the ellipsoid).

birocoles replied:
I removed the method `def V(...)`.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+        Calculates the coordinate transformation matrix
+        for a triaxial model. The columns of this matrix
+        are defined according to the unit vectors v1, v2
+        and v3 presented by Clark et al. (1986, p. 192).
```
The docstring format is:

    """
    One sentence short description.

    More text explaining what this is.
    Can include references.
    """

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+        Example:
+
+            >>> e = TriaxialEllipsoid(11, 3, 4, 3, 2, 1, 180, 90, 0)
+            >>> print e.V()
```
Don't print numpy arrays on doctests. This will break eventually for some Windows/Python version because of floating point precision. This test belongs on a test file and checking using something like `numpy.testing.assert_allclose`.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1720,3 +1720,433 @@ def keep(cell):
+
+        return matrix
+
+
```
The same comments applies to all classes.

leouieda commented on fatiando/tests/mesher/test_oblate_ellipsoid.py:
```diff
@@ -0,0 +1,49 @@
+from ...mesher import OblateEllipsoid
```
You can put all tests in the same file `test_ellipsoids.py` because they are very similar. That way it's easier to maintain.

birocoles replied:
All tests are in the same file now

leouieda commented on fatiando/tests/mesher/test_oblate_ellipsoid.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def test_oblate_ellipsoid_V_orthogonal():
+    a = 1000.*rand()
```
Never test using random values! There is no way to know what values causes the tests to fail and no way to reproduce the failing test. Use a specific value. If you want to test several values, use a list and a `for` loop.

leouieda replied:
These are great tests! Testing for consistency and using a known simple case. Perfect!

You can use the functions in `numpy.testing` instead of using `assert np.allclose`. They give much better reports when the tests fail. See https://github.com/fatiando/fatiando/blob/master/fatiando/gridder/tests/test_interpolation.py for example.

    import numpy.testing as npt
    npt.assert_allclose(array1, array2)

leouieda commented on fatiando/tests/mesher/test_oblate_ellipsoid.py:
```diff
@@ -0,0 +1,49 @@
+from numpy.random import rand
+
+
+def test_oblate_ellipsoid_copy():
```
Place a single line docstring explaining what this test is checking.

birocoles replied:
The line was included.

leouieda reviewed (changes requested):
@birocoles don't worry about the eqlayer test failing for now. I'll have a look at that later if it happens on other branches as well. I'll leave some comments here, mostly about the style and implementation ideas.

It would be good to get @DiegoTaka looking at this as well to learn about the code formatting.

eusoubrasileiro commented:
sorry @birocoles but it is so comforting seeing you being reviewed hahahaha you can call me names, I couldn't resist sorry @leouieda

birocoles commented:
@leouieda uhul! Let's work!

leouieda commented:
@birocoles @eusoubrasileiro that's the spirit!

eusoubrasileiro commented:
you are cute :+1:

leouieda commented:
@birocoles PR #374 fixes the PEL test that was failing in Windows 32bit. Merge changes in from master to get this change.

birocoles committed:
remove the extra space before ( and remove the function V

birocoles committed:
Fix the problem with the PEL test (#375)

* Fix failing tests due to numpy 1.12 update (#373)

A few tests started breaking with the new numpy update.
This PR fixes these tests.

The code for seismic.conv uses np.floor and np.ceil to get 
index number for numpy arrays. 
This now causes errors. 
Wrap both calls in `int(...)` to fix the problem.

* Fix failing pole reduction test in Windows 32bit (#374)

The pole reduction test for the PEL is failing in Windows 32bit
for some reason (probably some floating point rounding problem).
This PR fixes that test by adjusting the amount of regularization.

* Add AppVeyor badge to README (#372)

To inform that we test on Linux and Windows and provides
a quick link to the latest CI status.

birocoles committed:
revised classes without the method def V(...)

birocoles committed:
put all tests in the same file and review the doctests

birocoles committed:
move coord tran matrix, suscep tensor and orient angles to the ellipsoid classes

birocoles committed:
verify the triaxial coord transf matrix for spectific input values

leouieda committed:
Merge branch 'master' into mesher-ellipsoid

leouieda commented:
@birocoles please update your clone with `git pull` because I fixed some merge conflicts with master for the PEL tests that were failing.

birocoles committed:
separate the functions coord_transf_matrix and auxiliary_angles from the
ellipsoid classes

birocoles committed:
verify the orthogonalyty of the coord transf matrices

birocoles committed:
correct a signal in the v1 forming the coord transf matrix for oblate ellipsoids

birocoles committed:
test the function auxiliary_angles with known input

birocoles committed:
Merge branch 'master' into mesher-ellipsoid

birocoles committed:
import division and absolute_import from __future__

birocoles committed:
include tests for expected assertion errors

birocoles commented:
@leouieda, I tried to include all your request changes. During the review, I opted for creating three separated functions in `mesher.py`: `auxiliary_angles`, `coord_transf_matrix_triaxial` and `coord_transf_matrix_oblate`. I am not sure if it is the best way.

leouieda commented:
@birocoles the code is much better now and easier to understand. Great job 🥇 !

I see that the susceptibility calculation is still a bit strange. This is a hard design issue. I really liked your idea of a `utils` function for the suscep tensor but that probably won't work because it's a different coord trans matrix for each ellipsoid. So this is something tied to the ellipsoid and it should be in the class. That's why the coord trans matrix functions should be part of the classes as well. They just shouldn't be public methods. You can add a `_` to the beginning of a method name to indicate that it's meant for internal use by the class (`_coord_trans_matrix`).

One thing I didn't like is that you need to have a `"susceptibility tensor"` property in order to calculate the `susceptibility_tensor`. That seems strange. So maybe the `susceptibility_tensor` should be method name and something else for the physical property. Because you're not really giving the tensor right? It's the components needed to build it: 3 values and 3 angles. Maybe you could specify 2 properties: `"susceptibility principal"` and `"susceptibility angles"`. Then you can use both to calculate the tensor itself.

I'll leave some comments in the code as well.

This has really shown a light into how broken the `mesher` module really is. We need a reformulation of this. I'm thinking about a design and will probably post it somewhere and ask for input later.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+        # Coordinate transformation matrix
+        self.transf_matrix = coord_transf_matrix_triaxial(self.alpha,
+                                                          self.gamma,
+                                                          self.delta)
```
Why do you store both of these and then never use them again? Will they be needed in the forward modeling code?

birocoles replied:
Actually, only the `transf_matrix` is used in the forward modelling. I will remove the `self.` from `alpha`, `gamma` and `delta` .

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+
+
+class TriaxialEllipsoid(GeometricElement):
+
```
Empty line.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+    This code follows Clark et al. (1986) and defines the spatial
+    orientation of the ellipsoid by using three angles: strike,
+    rake and dip. These angles are commonly used to define the
+    orientation of geological structures (Allmendinger et al., 2012).
```
Nice! this is much better!

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+    * large_axis, intermediate_axis, small_axis: float
+        Semi-axes forming the ellipsoid (in m).
+    * strike, dip, rake
+        Orientation angles of the ellipsoid (in degrees).
```
Much better!

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+        Physical properties assigned to the ellipsoid.
+        Ex: ``props={'density':10,
+                     'remanent magnetization':[10, 25, 40],
+                     'susceptibility tensor':[0.562, 0.485, 0.25,
```
Is the `"remanent magnetization"` something that you'll need in the forward modeling later as well?

birocoles replied:
Yes. The computation of the resultant magnetization, including the self-demagnetization, depends on the remanent magnetization, local-geomagnetic field, susceptibility tensor (in matrix form) and a matrix containing the demagnetizing factors (depolarization tensor). The resultant magnetization, in turn, is used to calculate the magnetic field components.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+        >>> print(e)
+        x:1 | y:2 | z:3 | large_axis:6 | intermediate_axis:5 | small_axis:4 | \
+strike:10 | dip:20 | rake:30 | density:20 | remanent magnetization:[10, 25, 40\
+] | susceptibility tensor:[0.562, 0.485, 0.25, 90, 0, 0]
```
I've been reconsidering these long string representations of the objects. They're not really useful aside from the doctests, and even then they aren't pretty, specially when they get long like this one.

What I've been doing is printing individual attributes. For example: `print(e.x, e.y, e.z)`. That's not a very useful doctest as well because it doesn't check any calculation, just the setting of the attributes (which is pretty hard to mess up).

leouieda replied:
It would be better to test something like the susceptilbity tensor calculation.

birocoles replied:
**1. string representation**

What do you think about breaking the string representation into lines.?

**2. susceptibility tensor calculation**

The susceptibility tensor is a numpy array. The problem of creating a doctest illustrating the susceptibility tensor calculation is that it would be necessary to print a numpy array. What do you think?

leouieda replied:
> 1. string representation
> What do you think about breaking the string representation into lines?

I don't think that will help. It will still be ugly and not very useful. I prefer to just get rid of it. Take a look at what I've been doing in #378 (https://github.com/fatiando/fatiando/pull/378/files#diff-94a56fa30b5d3b4dc51db74b1bc2cae9). That PR ended up being too big and I'll break it into smaller parts later.

> 2. susceptibility tensor calculation
> The susceptibility tensor is a numpy array. The problem of creating a doctest illustrating the susceptibility tensor calculation is that it would be necessary to print a numpy array. What do you think?

Then we'd run into the precision problems we were having before. Hum. You're right. Maybe just leave the simple stuff in the doctests. This isn't a complicated class and there isn't much need for complex examples in the docstring. Use the test modules to test the tensor, as you've already been doing.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+
+    def __init__(self, x, y, z, large_axis, intermediate_axis, small_axis,
+                 strike, dip, rake, props=None):
+        GeometricElement.__init__(self, props)
```
This is the old style of calling the parent class `__init__`. The new way that's compatible with Python 3 is `super().__init__(props)`  (see the `inversion` package classes). You'll need to add the import `from future.builtins import super` if it's not there already.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+                 ('small_axis', self.small_axis), ('strike', self.strike),
+                 ('dip', self.dip), ('rake', self.rake)]
+        names = names + [(p, self.props[p]) for p in sorted(self.props)]
+        return ' | '.join('%s:%s' % (n, v) for n, v in names)
```
From the comment above, you could probably remove this. Then it's less code to maintain.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+        names = names + [(p, self.props[p]) for p in sorted(self.props)]
+        return ' | '.join('%s:%s' % (n, v) for n, v in names)
+
+    def susceptibility_tensor(self):
```
I'd prefer this to be a property. If you add `@property` above the `def` it will turn this method into a property. The  you can access the tensor without the `()` like `ellipsoid.susceptibility_tensor` and it calls the method behind the scenes.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+        '''
+
+        assert 'susceptibility tensor' in self.props, 'susceptibility tensor \
+is not in the dictionary of physical properties'
```
I think it would be better to `return None` if the required physical properties aren't present. Like `ellipsoid.susceptibility_tensor` -> `None` oh, I guess there is no suscep tensor.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+    aux1 = cos_rake/numpy.sqrt(1 - aux*aux)
+    aux2 = sin_dip*cos_rake
+
+    if aux1 > 1.:
```
This kind of check won't work if any of the inputs are numpy arrays but we don't need to worry about that now.

leouieda commented on fatiando/mesher.py:
```diff
@@ -1721,3 +1721,594 @@ def keep(cell):
+        return suscep_tensor
+
+
+def auxiliary_angles(strike, dip, rake):
```
This method is very specific to the ellipsoid code so it should be private (add a `_` to the beginning of the name).

birocoles committed:
include super() for calling the parent class and set the susceptibility_tensor as a @property

birocoles commented:
@leouieda , I liked the idea of splitting the property `susceptibility tensor` into `principal susceptibilities` and `susceptibility angles`. I will do it in the next commit.

birocoles committed:
split 'susceptibility tensor' into 'principal susceptibilities' and 'susceptibility angles'

birocoles committed:
corret the import mesher

leouieda committed:
Merge branch 'master' into mesher-ellipsoid

Conflicts:
	fatiando/mesher/mesh.py

leouieda commented:
@birocoles I updated the PR with the latest master and the changes from #383 . The ellipsoid code is in `fatiando/mesher/geometry.py` and the tests in `fatiando/mesher/tests/test_geometry.py`. The rest is all the same and all tests pass.
